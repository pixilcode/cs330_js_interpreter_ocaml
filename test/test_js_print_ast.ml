open Js_interpreter

let run_acorn js =
  (* write the js to a temp file *)
  let temp_js_file = Filename.temp_file "acorn" ".js" in
  let temp_json_file = Filename.temp_file "acorn" ".json" in
  let oc = open_out temp_js_file in
  output_string oc js;
  close_out oc;

  (* run acorn on the temp file *)
  ignore (Sys.command ("acorn --ecma2024 " ^ temp_js_file ^ " > " ^ temp_json_file) : int);

  (* read the result *)
  let ic = open_in_bin temp_json_file in
  let json = really_input_string ic (in_channel_length ic) in
  close_in ic;

  (* delete the temp files *)
  Sys.remove temp_js_file;
  Sys.remove temp_json_file;

  json

(* run a test and print the result *)
let run_test (name, js) =
  print_endline ("--- " ^ name ^ " ---");

  let json = run_acorn js in
  let ast = Ast.from_json_string json in
  let s_expr = S_expr.from_ast ast in

  print_endline (S_expr.pretty_print s_expr);
  print_newline ()

let tests = [
  ("simple_number", "1");
  ("unary", "!true");
  ("arithmetic", "1 + 2 - 3 * 4 / 5");
  ("relational", "1 < 2 == true");
  ("logical", "true && false || false")
]

(* iterate over all the tests and run them *)
let () = List.iter run_test tests