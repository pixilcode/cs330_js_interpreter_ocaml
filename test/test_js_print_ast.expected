--- simple_number ---
(number 1)

--- unary ---
(unary ! (boolean true))

--- arithmetic ---
(arithmetic - (arithmetic + (number 1) (number 2)) (arithmetic / (arithmetic * (number 3) (number 4)) (number 5)))

--- relational ---
(relational == (relational < (number 1) (number 2)) (boolean true))

--- logical ---
(logical || (logical && (boolean true) (boolean false)) (boolean false))

